### DISCUSSION TITLE
Poor responsiveness and high CPU load

### DISCUSSION THREAD

**Discussion by thinkbig1979** [Budibase/budibase#13928] [Help]
I've built a CRM in self-hosted Budibase (docker compose), 
Version: 2.28.4
Using a local Postgres DB in the stack as primary data provider.
My users are complaining - and rightly so - about poor responsiveness of the system.

Every time you go to a different page in the CRM, the the app-service and CouchDB-service containers max out the CPU for about 10 seconds while the page loads. The logs show tons of calls to couchDB when changing pages, even if I'm only loading a single view of a single datasource, e.g. my contacts table.
 
Looking at the logs, I cannot tell if Redis is being used at all.

Any suggestions?

EDIT: corrected the containers that are maxing CPU. It wasn't the worker container, but the app container.

**Comment by aptkingston** [comment]
Hey @thinkbig1979. Does your table contain large amounts of heavily related data? The only thing I can think of is that Budibase is having to do a large amount of joins when fetching data due to how our relationship system works.

Using a single table on a single screen should be almost instant in normal circumstances.

I'm also assuming that querying similar data directly from PG doesn't have any performance issues? Just to rule that out. It's likely the Budibase relationship system that's the bottleneck here if you have heavily related data with large numbers of row in every table.

If you could shed some light on the rough schema and size of your tables that would help identify the problem. Thanks!

**Comment by thinkbig1979** [comment]
Sure, focusing on the contacts table: Here's the postgres schema. I also have a formula column that concatenates 3 fields into the full name column, but only 2 relations to other tables. Total entries around the 500 mark, so really nothing special.
![image](https://github.com/Budibase/budibase/assets/18327273/985d07c0-6a36-4478-b583-633d585e332b)

I do have about 5 other tables that tie into this one, so were those tables have foreign keys in the contacts table, so about 7 relationships in all. 


Some additional info:
If I pull up my data view of the contacts table, it takes a few seconds to render the entire table, including all the relational fields.
If I go to my contacts page in the CRM app, again, just a single view of that table with nothing else on that page (on initial load), it takes 20-30 seconds, with significantly more going on in the logs of the app and couchDB services than when I view the table. I do have that page set up to display certain relationships such as employer data in cards below the contacts table when I select a contact that has that info populated. So all all I am tapping into various data sources on that page. Perhaps Budibase is trying to pre-hydrate all employer data on all contact records or something, but at the moment we only have employer data filled in for about 20 contacts, so you wouldn't think it was that  much to do that it should take 30 seconds and max out the CPU for quite a while.  

Update:
If I go into Metabase, and create a table that joins everything connected to my contacts table and pulls in **all columns** from **all** joined tables, I end up with an SQL statement of over 130 lines with 7 joins. That query takes 5 seconds to run on the same machine, and maxes out the CPU during most of that time. This seems to match roughly what happens when I view the contacts table in the data view on a clean load of Budibase.
